packages/crypto/src/encryption.ts

```diff
@@ -1,26 +1,93 @@
-import SHA256 from 'crypto-js/sha256';
-import AES from 'crypto-js/aes';
-import Utf8 from 'crypto-js/enc-utf8';
-
-const STRETCH_ROUNDS = 5000;
-
-// In cryptography, key stretching techniques are used to make brute-force attacks more difficult.
-// The idea is to "stretch" the key (or in this case, the password) by applying a computationally intensive
-// operation to it multiple times. This makes each attempt to guess the password more time-consuming,
-// which in turn makes brute-force attacks less feasible.
-export const repeatedHash = (message: string) => {
-  let hash = SHA256(message);
-  for (let i = 0; i < STRETCH_ROUNDS; i++) {
-    hash = SHA256(hash.toString());
+import { Base64Str, base64ToUint8Array, uint8ArrayToBase64 } from 'penumbra-types';
+
+// Salts & initialization vectors should be stored
+export const randomBase64str = (): Base64Str => {
+  const uintArr = crypto.getRandomValues(new Uint8Array(16)); // 128 bits
+  return uint8ArrayToBase64(uintArr);
+};
+
+export const hashPassword = async (password: string, salt: Base64Str): Promise<JsonWebKey> => {
+  const enc = new TextEncoder();
+  const importedKey = await crypto.subtle.importKey(
+    'raw',
+    enc.encode(password),
+    { name: 'PBKDF2' },
+    false,
+    ['deriveBits', 'deriveKey'],
+  );
+
+  const derivedKey = await crypto.subtle.deriveKey(
+    {
+      name: 'PBKDF2',
+      salt: base64ToUint8Array(salt),
+      iterations: 210_000,
+      hash: 'SHA-512',
+    },
+    importedKey,
+    { name: 'AES-GCM', length: 256 },
+    true,
+    ['encrypt', 'decrypt'],
+  );
+
+  return crypto.subtle.exportKey('jwk', derivedKey);
+};
+
+export const isPassword = async (
+  password: string,
+  salt: Base64Str,
+  encryptedSeedPhrase: Base64Str,
+  initializationVector: Base64Str,
+): Promise<boolean> => {
+  try {
+    const key = await hashPassword(password, salt);
+    await decrypt(encryptedSeedPhrase, initializationVector, key);
+    return true;
+  } catch (error) {
+    return false;
   }
-  return hash.toString();
 };
 
-export const encrypt = (message: string, key: string) => {
-  return AES.encrypt(message, key).toString();
+export const encrypt = async (
+  message: string,
+  initializationVector: Base64Str,
```

Comment:
Being able to choose the IV is a bit of a footgun.

It seems that the rest of the code we have using this does a reasonable thing and generates a random IV, but we could instead simplify that code by having this encryption function generate a random IV and prepend it to the ciphertext, and then decryption would grab the IV from the ciphertext + IV combo directly.

This would be opaque, and just an internal detail of how this little library encodes ciphertexts.